agent/dns_test.go

```diff
@@ -3832,6 +3832,45 @@ func TestDNS_Recurse_Truncation(t *testing.T) {
 	}
 }
 
+func TestDNS_Recursor_StrategyRandom(t *testing.T) {
+	configuredRecursors := []string{"1.1.1.1", "8.8.4.4", "8.8.8.8"}
+	recursorStrategy := agentdns.RecursorStrategy("random")
+
+	retries := func() *retry.Counter {
+		return &retry.Counter{Count: 5}
+	}
+
+	retry.RunWith(retries(), t, func(r *retry.R) {
+		recursorsToQuery := make([]string, 0)
+		for _, idx := range recursorStrategy.Indexes(len(configuredRecursors)) {
+			recursorsToQuery = append(recursorsToQuery, configuredRecursors[idx])
+		}
+
+		// Ensure the slices contain the same elements
+		require.ElementsMatch(t, configuredRecursors, recursorsToQuery)
+
+		if reflect.DeepEqual(configuredRecursors, recursorsToQuery) {
+			// Error if the elements are in the same order, and retry generating
+			// random recursor list
+			r.Fatal("dns recursor order is not randomized.")
+		}
```

Comment:
I think this also works:

```go
require.Equal(r, configuredRecursors, recursorsToQuery)
```

Reply:
I initially tried this approach, but when running the test under the debugger I did not see the retry loop being executed if the recursor order was equal. Using `require.Equal(r, configuredRecursors, recursorToQuery)` seems to cause the test to fail silently, and appear to `go test` as if the test passed, whereas `r.Fatal()` correctly triggers the retry loop.

Reply:
Oh, that's concerning. I made some changes to retry recently. I'll double check how it is working.

Reply:
Oh, right of course `require.Equal` is not at all what we want! It should be `require.NotEqual` to match this behaviour.

But I think what you have here is good too

Reply:
You're right, I should've been using `require.NotEqual()`. 🤦‍♂️ I just updated the code to use this function and the retry is working as expected.